c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -0,0 +1,370 @@
+    //mir: &'a Body<'tcx>,
+    //span_index: SpanIndex<Location>,
+    hir_rewrites: Vec<(Span, Rewrite)>,
+}
+
+impl<'a, 'tcx> HirTyVisitor<'a, 'tcx> {
+    fn handle_ty(&mut self, rw_lty: RwLTy<'tcx>, hir_ty: &hir::Ty<'tcx>) {
+        if rw_lty.label.ty_desc.is_none() && !rw_lty.label.descendant_has_rewrite {
+            // No rewrites here or in any descendant of this HIR node.
+            return;
+        }
+
+        let hir_args = match deconstruct_hir_ty(rw_lty.ty, hir_ty) {
+            Some(x) => x,
+            None => {
+                // `hir_ty` doesn't have the expected structure (for example, we expected a type
+                // like `*mut T`, but it's actually an alias `MyPtr`), so we can't rewrite inside
+                // it.  Instead, we discard it completely and pretty-print `rw_lty` (with rewrites
+                // applied).
+                let ty = mk_rewritten_ty(self.rw_lcx, rw_lty);
+                let printer = FmtPrinter::new(*self.rw_lcx, Namespace::TypeNS);
+                let s = ty.print(printer).unwrap().into_buffer();
+                self.hir_rewrites.push((hir_ty.span, Rewrite::PrintTy(s)));
+                return;
+            }
+        };
+
+        if let Some((own, qty)) = rw_lty.label.ty_desc {
+            assert_eq!(hir_args.len(), 1);
+            let mut rw = Rewrite::Sub(0, hir_args[0].span);
+
+            if own == Ownership::Cell {
+                rw = Rewrite::TyCtor("core::cell::Cell".into(), vec![rw]);
+            }
+
+            rw = match qty {
+                Quantity::Single => rw,
+                Quantity::Slice => Rewrite::TySlice(Box::new(rw)),
+                // TODO: This should generate `OffsetPtr<T>` rather than `&[T]`, but `OffsetPtr` is
+                // NYI
+                Quantity::OffsetPtr => Rewrite::TySlice(Box::new(rw)),
+            };
+
+            rw = match own {
+                Ownership::Raw => Rewrite::TyPtr(Box::new(rw), Mutability::Not),
+                Ownership::RawMut => Rewrite::TyPtr(Box::new(rw), Mutability::Mut),
+                Ownership::Imm => Rewrite::TyRef(Box::new(rw), Mutability::Not),
+                Ownership::Cell => Rewrite::TyRef(Box::new(rw), Mutability::Not),
+                Ownership::Mut => Rewrite::TyRef(Box::new(rw), Mutability::Mut),
+                Ownership::Rc => todo!(),
+                Ownership::Box => todo!(),
+            };
+
+            self.hir_rewrites.push((hir_ty.span, rw));
+        }
+
+        if rw_lty.label.descendant_has_rewrite {
+            for (&arg_rw_lty, arg_hir_ty) in rw_lty.args.iter().zip(hir_args.into_iter()) {
+                self.handle_ty(arg_rw_lty, arg_hir_ty);
+            }
+        }
+    }
+}
+
+impl<'a, 'tcx> intravisit::Visitor<'tcx> for HirTyVisitor<'a, 'tcx> {
+    type NestedFilter = nested_filter::OnlyBodies;
+
+    fn nested_visit_map(&mut self) -> Self::Map {
+        self.acx.tcx().hir()
+    }
+}
+
+pub fn gen_ty_rewrites<'tcx>(
```

Comment:
is this only for function signatures and not function bodies?

Reply:
Yes, function signatures only for now.  There are TODOs below for "update let bindings" and "update cast RHS types", which would cover types that appear inside function bodies.  There's currently no TODO comment about struct fields, since those weren't supported when I wrote this code, but I'll add one.